engines/lab/dispman.cpp

```diff
@@ -0,0 +1,1022 @@
+		invButtonList->push_back(e->createButton(234, y, 6, 'b', invImages[6], invImages[7]));
+		invButtonList->push_back(e->createButton(266, y, 7, 'f', invImages[8], invImages[9]));
+	}
+
+	delete invFile;
+}
+
+void DisplayMan::setPen(byte penNum) {
+	_curPen = penNum;
+}
+
+void DisplayMan::rectFill(Common::Rect fillRect) {
+	int width = fillRect.width() + 1;
+	int height = fillRect.height() + 1;
+
+	if (fillRect.left + width > _screenWidth)
+		width = _screenWidth - fillRect.left;
+
+	if (fillRect.top + height > _screenHeight)
+		height = _screenHeight - fillRect.top;
+
+	if ((width > 0) && (height > 0)) {
+		char *d = (char *)getCurrentDrawingBuffer() + fillRect.top * _screenWidth + fillRect.left;
+
+		while (height-- > 0) {
+			char *dd = d;
+			int ww = width;
+
+			while (ww-- > 0) {
+				*dd++ = _curPen;
+			}
+
+			d += _screenWidth;
+		}
+	}
+}
+
+void DisplayMan::rectFill(uint16 x1, uint16 y1, uint16 x2, uint16 y2) {
+	rectFill(Common::Rect(x1, y1, x2, y2));
+}
+
+void DisplayMan::rectFillScaled(uint16 x1, uint16 y1, uint16 x2, uint16 y2) {
+	rectFill(_vm->_utils->vgaRectScale(x1, y1, x2, y2));
+}
+
+void DisplayMan::drawVLine(uint16 x, uint16 y1, uint16 y2) {
+	rectFill(x, y1, x, y2);
+}
+
+void DisplayMan::drawHLine(uint16 x1, uint16 y, uint16 x2) {
+	rectFill(x1, y, x2, y);
+}
+
+void DisplayMan::screenUpdate() {
+	g_system->copyRectToScreen(_displayBuffer, _screenWidth, 0, 0, _screenWidth, _screenHeight);
+	g_system->updateScreen();
+
+	_vm->_event->processInput();
+}
+
+void DisplayMan::createScreen(bool hiRes) {
+	if (hiRes) {
+		_screenWidth  = 640;
+		_screenHeight = 480;
+	} else {
+		_screenWidth  = 320;
+		_screenHeight = 200;
+	}
+	_screenBytesPerPage = _screenWidth * _screenHeight;
+
+	if (_displayBuffer)
+		delete[] _displayBuffer;
+	_displayBuffer = new byte[_screenBytesPerPage];
+}
+
+void DisplayMan::setAmigaPal(uint16 *pal, uint16 numColors) {
+	byte vgaPal[16 * 3];
+	uint16 vgaIdx = 0;
+
+	if (numColors > 16)
+		numColors = 16;
+
+	for (uint16 i = 0; i < numColors; i++) {
+		vgaPal[vgaIdx++] = (byte)(((pal[i] & 0xf00) >> 8) << 2);
+		vgaPal[vgaIdx++] = (byte)(((pal[i] & 0x0f0) >> 4) << 2);
+		vgaPal[vgaIdx++] = (byte)(((pal[i] & 0x00f)) << 2);
+	}
+
+	writeColorRegs(vgaPal, 0, 16);
+	_vm->waitTOF();
+}
+
+void DisplayMan::writeColorRegs(byte *buf, uint16 first, uint16 numReg) {
+	byte tmp[256 * 3];
+
+	for (int i = 0; i < 256 * 3; i++)
+		tmp[i] = buf[i] * 4;
+
+	g_system->getPaletteManager()->setPalette(tmp, first, numReg);
+	memcpy(&(_curvgapal[first * 3]), buf, numReg * 3);
+}
+
+void DisplayMan::setPalette(void *newPal, uint16 numColors) {
+	if (memcmp(newPal, _curvgapal, numColors * 3) != 0)
+		writeColorRegs((byte *)newPal, 0, numColors);
+}
+
+byte *DisplayMan::getCurrentDrawingBuffer() {
+	if (_currentDisplayBuffer)
+		return _currentDisplayBuffer;
+
+	return _displayBuffer;
+}
+
+void DisplayMan::overlayRect(uint16 penColor, uint16 x1, uint16 y1, uint16 x2, uint16 y2) {
```

Comment:
My understanding is that this tries to draw all filled pieces of a checkerboard pattern (no promises on that though). I think we should either adapt the function name or make the code more obvious as to what it is doing.